python/ray/tune/integration/kubernetes.py

```diff
@@ -0,0 +1,143 @@
+        self.worker_ip = worker_ip
+        self.worker_node = self._get_kubernetes_node_by_ip(worker_ip)
+
+    def _get_kubernetes_node_by_ip(self, node_ip):
+        """Return node name by internal or external IP"""
+        kubernetes.config.load_incluster_config()
+        api = kubernetes.client.CoreV1Api()
+        pods = api.list_namespaced_pod(self._namespace)
+        for pod in pods.items:
+            if pod.status.host_ip == node_ip or \
+               pod.status.pod_ip == node_ip:
+                return pod.metadata.name
+
+        logger.error(
+            "Could not find Kubernetes pod name for IP {}".format(node_ip))
+        return None
+
+    @property
+    def _remote_path(self):
+        return (self.worker_node, self._remote_dir)
+
+
+class KubernetesSyncClient(SyncClient):
+    """KubernetesSyncClient to be used by KubernetesSyncer.
+
+    This client takes care of executing the synchronization
+    commands for Kubernetes clients. In its ``sync_down`` and
+    ``sync_up`` commands, it expects tuples for the source
+    and target, respectively, for compatibility with the
+    KubernetesCommandRunner.
+
+    Args:
+        namespace (str): Namespace in which the pods live.
+        process_runner: How commands should be called.
+            Defaults to ``subprocess``.
+
+    """
+
+    def __init__(self, namespace, process_runner=subprocess):
+        self.namespace = namespace
+        self._process_runner = process_runner
+        self._command_runners = {}
+
+    def _create_command_runner(self, node_id):
+        """Create a command runner for one Kubernetes node"""
+        return KubernetesCommandRunner(
+            log_prefix="KubernetesSyncClient: {}:".format(node_id),
+            namespace=self.namespace,
+            node_id=node_id,
+            auth_config=None,
+            process_runner=self._process_runner)
+
+    def _get_command_runner(self, node_id):
+        """Create command runner if it doesn't exist"""
+        if node_id not in self._command_runners:
+            command_runner = self._create_command_runner(node_id)
+            self._command_runners[node_id] = command_runner
+        return self._command_runners[node_id]
+
+    def sync_up(self, source, target):
+        """Here target is a tuple (target_node, target_dir)"""
+        target_node, target_dir = target
+
+        command_runner = self._get_command_runner(target_node)
+        command_runner.run_rsync_up(source, target_dir)
+        return True
+
+    def sync_down(self, source, target):
+        """Here source is a tuple (source_node, source_dir)"""
+        source_node, source_dir = source
+
+        command_runner = self._get_command_runner(source_node)
+        command_runner.run_rsync_down(source_dir, target)
+        return True
+
+    def delete(self, target):
+        """No delete function because it is only used by
+        the KubernetesSyncer, which doesn't call delete."""
+        return True
```

Comment:
what does this mean?

Reply:
(I guess why doesn't the kubernetesSyncer call delete?)

Reply:
Generally ``Syncer``s never seem to call delete. They don't have an interface for that and don't use it internally. ``SyncClient``s (``FunctionBasedClient`` and ``CommandBasedClient``) are sometimes used by other objects which can call ``delete()`` (e.g. the ``DurableTrainable``), but this isn't the case for the ``KubernetesSyncClient``, which is only instantiated by the new ``KubernetesSyncClient``. If we extend usage in the future, we can (and should) introduce that functionality.

Reply:
Make sense!